astropy/tests/helper.py

```diff
@@ -412,11 +413,16 @@ def generic_recursive_equality_test(a, b, class_history):
     Check if the attributes of a and b are equal. Then,
     check if the attributes of the attributes are equal.
     """
-    dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    if PYTHON_LT_3_11:
+        dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    else:
+        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
+        # with required argument(s) are passed to this function.
+        # For a class that has no instance __dict__ and no __slots__, the default state is None
+        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__() or dict()
```

Comment:
```suggestion
        # For a class that has no instance __dict__ and no __slots__, the default state is None
        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__() or dict()
        # For a class that has __slots__ and no instance __dict__, the default state is a tuple whose
        # first item is None and whose second item is a dictionary mapping slot names to slot values.
        if insinstance(dict_a, tuple):
            dict_a = dict_a[1]
```

Reply:
> For a class with `__slots__` or no `__dict__`, the default state is not a

Hmm, the Python docs say
"For a class that has **no** instance `__dict__` and **no** `__slots__`, the default state is `None`.".
But in the cosmology example we have
```python
>>> print(FLRW.H0.fvalidate)
<function _validate_to_scalar at 0x126565a80>
>>> hasattr(FLRW.H0.fvalidate, '__slots__')
False
>>> print(FLRW.H0.fvalidate.__dict__)
{}
>>> print(FLRW.H0.fvalidate.__getstate__())
None
```
The above case would still break in the (so far hypothetical) case of something that has no `__dict__`.

Reply:
I've modified your suggestion to account for that corner case.
In the sense of "if there is no usable `__getstate__()` return, use `__dict__` that probably does match the pre-3.11 idea better.

Reply:
@pllim if @eerovaher agrees to the changed suggestion, let's test that and we should be done (and feed it back into 5.1 and main).

Reply:
Your edit to my suggestion looks good and if the tests succeed then it would indeed be a good idea to apply that patch to the other branches.